File: practice/algorithms/implementation/sock_merchant/SockMerchant.scala
import scala.io.Source

object SockMerchant extends App {
  val lines = Source.stdin.getLines().toList
  val socks = lines(1).split(" ").map(_.toInt)
  val sockGroups = socks.groupBy(c => c)
  val sockGroupPairs = sockGroups.map(group => group._2.length / 2)
  val pairs = sockGroupPairs.sum

  println(pairs)
}

File: practice/algorithms/implementation/sock_merchant/sock_merchant.rb
n = gets.to_i
socks = gets.split(" ")
sock_groups = Hash.new
socks.each do |s|
  if !sock_groups.include?(s)
    sock_groups[s] = 1
  else
    count = sock_groups[s]
    sock_groups[s] = count + 1
  end
end

pairs = 0
sock_groups.values.each { |v| pairs += v / 2}

puts pairs